File: Advent Code 2024/Day09.js
const fs = require('fs');

function parseDiskMap(diskMap) {
    const result = [];
    let fileID = 0;

    for (let i = 0; i < diskMap.length; i += 2) {
        const fileLen = parseInt(diskMap[i]);
        const freeSpace = parseInt(diskMap[i + 1]);

        for (let j = 0; j < fileLen; j++) result.push(fileID);
        for (let j = 0; j < freeSpace; j++) result.push('.');

        fileID++;
    }

    return result;
}

function compactDisk(disk) {
    const compacted = disk.filter((block) => block !== '.'); // Remove free spaces
    while (compacted.length < disk.length) compacted.push('.');
    return compacted;
}

function calculateChecksum(disk) {
    let checksum = 0;
    disk.forEach((block, position) => {
        if (block !== '.') {
            checksum += position * parseInt(block);
        }
    });
    return checksum;
}

function main() {
    const diskMap = fs.readFileSync("Lists/data9.txt", "utf-8").trim();

    let disk = parseDiskMap(diskMap);
    console.log("Initial Disk:", disk.join(''));

    disk = compactDisk(disk);
    console.log("Compacted Disk:", disk.join(''));

    const checksum = calculateChecksum(disk);
    console.log("Checksum:", checksum);
}

main();

File: Advent_Of_Code_2024/Day09.py
def parse_disk_map(disk_map):
    if len(disk_map) % 2 != 0:
        raise ValueError("Disk map length must be even.")

    result = []
    file_id = 0
    for i in range(0, len(disk_map), 2):
        file_len = int(disk_map[i])
        free_space = int(disk_map[i + 1])
        result.extend([file_id] * file_len)  # Add file blocks
        result.extend(['.'] * free_space)  # Add free space
        file_id += 1
    return result


def compact_disk(disk):
    compacted = []
    for block in disk:
        if block != '.':
            compacted.append(block)
    compacted.extend(['.'] * (len(disk) - len(compacted)))
    return compacted


def calculate_checksum(disk):
    checksum = 0
    for position, block in enumerate(disk):
        if block != '.':
            checksum += position * int(block)
    return checksum


def main():
    try:
        with open("Lists/data9.txt", "r") as file:
            disk_map = file.read().strip()

        if not disk_map:
            raise ValueError("Disk map cannot be empty.")

        disk = parse_disk_map(disk_map)
        print("Initial Disk:", ''.join(map(str, disk)))

        disk = compact_disk(disk)
        print("Compacted Disk:", ''.join(map(str, disk)))

        checksum = calculate_checksum(disk)
        print("Checksum:", checksum)
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()
